sklearn/tree/_criterion.pyx

```diff
@@ -1325,3 +1327,122 @@ cdef class FriedmanMSE(MSE):
+        i.e. the impurity of samples[start:end]. The smaller the impurity the
+        better.
+        """
+        cdef double* sum_total = self.sum_total
+        cdef double y_mean = 0.
+        cdef double impurity = 0.
+        cdef SIZE_t i, p, k
+        cdef DOUBLE_t w = 1.0
+
+        for k in range(self.n_outputs):
+            y_mean = sum_total[k] / self.weighted_n_node_samples
+            for p in range(self.start, self.end):
+                i = self.samples[p]
+
+                if self.sample_weight != NULL:
+                    w = self.sample_weight[i]
+
+                impurity += w * xlogy(self.y[i, k], self.y[i, k]/y_mean)
+
+        return impurity / (self.weighted_n_node_samples * self.n_outputs)
+
+    cdef double proxy_impurity_improvement(self) nogil:
+        """Compute a proxy of the impurity reduction.
+
+        This method is used to speed up the search for the best split.
+        It is a proxy quantity such that the split that maximizes this value
+        also maximizes the impurity improvement. It neglects all constant terms
+        of the impurity decrease for a given split.
+
+        The absolute impurity improvement is only computed by the
+        impurity_improvement method once the best split has been found.
+
+        Poisson proxy is:
+            - 1/n * sum(y_i * log(y_pred)) = -mean(y_i) * log(mean(y_i))
+        """
+        cdef SIZE_t k
+        cdef double proxy_impurity_left = 0.0
+        cdef double proxy_impurity_right = 0.0
+        cdef double y_mean_left = 0.
+        cdef double y_mean_right = 0.
+
+        for k in range(self.n_outputs):
+            y_mean_left = self.sum_left[k] / self.weighted_n_left
+            y_mean_right = self.sum_right[k] / self.weighted_n_right
+            if (y_mean_left <= 0) or (y_mean_right <= 0):
+                # Poisson loss does not allow non-positive predictions. We
+                # therefore forbid splits that have child nodes withs
+                # sum(y_i) <= 0.
+                return -INFINITY
+            else:
+                proxy_impurity_left += y_mean_left * log(y_mean_left)
+                proxy_impurity_right += y_mean_right * log(y_mean_right)
+
+        return - proxy_impurity_left - proxy_impurity_right
+
+    cdef void children_impurity(self, double* impurity_left,
+                                double* impurity_right) nogil:
+        """Evaluate the impurity in children nodes.
+
+        i.e. the impurity of the left child (samples[start:pos]) and the
+        impurity the right child (samples[pos:end]) for Poisson.
+        """
+        cdef const DOUBLE_t[:, ::1] y = self.y
+
+        cdef SIZE_t start = self.start
+        cdef SIZE_t pos = self.pos
+        cdef SIZE_t end = self.end
+
+        cdef SIZE_t i, p, k
+        cdef DOUBLE_t y_mean = 0.
+        cdef DOUBLE_t w = 1.0
+
+        impurity_left[0] = 0.
+        for k in range(self.n_outputs):
```

Comment:
any chance we can collect some statistics only once for the left node, and then get it for free for the right node as done for MSE? (Instead of looping twice?)

Reply:
See the `FIXME` comments. Omitting constant terms, it would suffice to compute `y_mean`. The full Poisson deviance needs `y_true_i * log(y_true_i)` for every sample `i`. Maybe, one could define this as as statistics like/instead of `sq_sum_total`? 

Since I wanted a small PR, I did not consider this - and also, I was a bit frightened by the implementation of `cdef class MAE`:smirk: